kube-runtime/src/controller/mod.rs

```diff
@@ -162,19 +166,30 @@ where
     QueueStream: TryStream<Ok = ObjectRef<K>>,
     QueueStream::Error: std::error::Error + 'static,
 {
+    let (scheduler_shutdown_tx, scheduler_shutdown_rx) = channel::oneshot::channel();
     let err_context = context.clone();
     let (scheduler_tx, scheduler_rx) = channel::mpsc::channel::<ScheduleRequest<ObjectRef<K>>>(100);
     // Create a stream of ObjectRefs that need to be reconciled
     trystream_try_via(
         // input: stream combining scheduled tasks and user specified inputs event
         Box::pin(stream::select(
             // 1. inputs from users queue stream
-            queue.context(QueueError).map_ok(|obj_ref| ScheduleRequest {
-                message: obj_ref,
-                run_at: Instant::now() + Duration::from_millis(1),
-            }),
+            queue
+                .context(QueueError)
+                .map_ok(|obj_ref| ScheduleRequest {
+                    message: obj_ref,
+                    run_at: Instant::now() + Duration::from_millis(1),
+                })
+                .on_complete(async move {
+                    // On error: scheduler has already been shut down and there is nothing for us to do
+                    let _ = scheduler_shutdown_tx.send(());
+                    tracing::debug!("applier queue terminated, starting graceful shutdown")
+                }),
```

Comment:
oh boy, the wild flow in this function continues to grow

Reply:
Yeah, I know, it's an abomination :/